[WebGPU] Support VideoFrame as the source of GPUExternalTexture

mwyrzykowski opened the pull request:
#### ededdcafd115f57ac989263d6247eb55ef7cd614
<pre>
[WebGPU] Support VideoFrame as the source of GPUExternalTexture
<a href="https://bugs.webkit.org/show_bug.cgi?id=256875">https://bugs.webkit.org/show_bug.cgi?id=256875</a>
&lt;radar://109442655&gt;

Reviewed by Myles C. Maxfield.

Support VideoFrame as the source of a GPUExternalTexture.

* Source/WebCore/Modules/WebGPU/GPUExternalTextureDescriptor.h:
(WebCore::GPUExternalTextureDescriptor::mediaIdentifierForSource):
(WebCore::GPUExternalTextureDescriptor::convertToBacking const):
* Source/WebCore/Modules/WebGPU/GPUExternalTextureDescriptor.idl:
* Source/WebCore/PAL/pal/graphics/WebGPU/WebGPUExternalTextureDescriptor.h:
(PAL::WebGPU::HTMLVideoElementIdentifier::HTMLVideoElementIdentifier):
(PAL::WebGPU::WebCodecsVideoFrameIdentifier::WebCodecsVideoFrameIdentifier):
Support either HTMLVideoElement or WebCodecsVideoFrame as the source of a GPUExternalTexture.

* Source/WebCore/platform/VideoFrame.h:
(WebCore::VideoFrame::resourceIdentifierVector const):
Default implementation returns and empty byte vector.

* Source/WebCore/platform/VideoPixelFormat.cpp:
(WebCore::convertVideoFramePixelFormat):
Allow kCVPixelFormatType_Lossless_420YpCbCr8BiPlanarVideoRange formats.

* Source/WebGPU/WebGPU/ExternalTexture.h:
(WebGPU::ExternalTexture::pixelBuffer const):
Retain the CVPixelBuffer which wasn&apos;t previously occurring.

* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.cpp:
(WebKit::GPUConnectionToWebProcess::createRemoteGPU):
Update the lambda to work with HTMLVideoElements and VideoFrames as inputs.

* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteDevice.cpp:
(WebKit::populateConvertedDescriptor):
(WebKit::RemoteDevice::importExternalTexture):
Support HTMLVideoElements and VideoFrames as inputs.

* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteDevice.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteGPU.h:

* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/WebGPU/WebGPUExternalTextureDescriptor.cpp:
(WebKit::WebGPU::ConvertToBackingContext::convertToBacking):
(WebKit::WebGPU::ConvertFromBackingContext::convertFromBacking):
PAL and WebCore representations are nearly identical.

* Source/WebKit/Shared/WebGPU/WebGPUExternalTextureDescriptor.h:
* Source/WebKit/Shared/WebGPU/WebGPUExternalTextureDescriptor.serialization.in:
* Source/WebKit/WebProcess/GPU/media/RemoteVideoFrameProxy.cpp:
(WebKit::RemoteVideoFrameProxy::resourceIdentifierVector const):
Encode the readRefrence into a byte-vector.

* Source/WebKit/WebProcess/GPU/media/RemoteVideoFrameProxy.h:

Canonical link: <a href="https://commits.webkit.org/264847@main">https://commits.webkit.org/264847@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/413756e7a0b12b9939726bbde93f0966d55832ad

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/8846 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/9134 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/9352 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/10499 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/8826 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/8854 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/11121 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/9101 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/11690 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/8992 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/9989 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/7825 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/10657 "Built successfully") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/8971 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/7287 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/8090 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/15584 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/8395 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/8238 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/11571 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/8723 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/7109 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/7979 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2143 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/12190 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/8471 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

jyavenard commented on Source/WebKit/GPUProcess/graphics/WebGPU/RemoteDevice.cpp:
```diff
@@ -152,24 +153,36 @@ void RemoteDevice::createSampler(const WebGPU::SamplerDescriptor& descriptor, We
+    }, [] (PAL::WebGPU::WebCodecsVideoFrameIdentifier i) -> MediaPlayerOrVideoFrameResult {
+#if PLATFORM(COCOA)
+        RemoteVideoFrameReference r;
+        memcpy(reinterpret_cast<void*>(&r), &i.value[0], i.value.size());
```
can we have a strong release assert that i.value.size() is the same as sizeof(RemoteVideoFrameReference).

Is there no other way to create a RemoteVideoFrameReference? can't we add a constructor instead?

mwyrzykowski replied:
I think I can use identifier and version instead to create the RemoteVideoFrameReference, I will try that

jyavenard commented on Source/WebKit/GPUProcess/GPUConnectionToWebProcess.cpp:
```diff
@@ -647,8 +647,14 @@ void GPUConnectionToWebProcess::createRemoteGPU(WebGPUIdentifier identifier, Ren
-        return IPC::ScopedActiveMessageReceiveQueue { RemoteGPU::create([this] (MediaPlayerIdentifier identifier, Function<void(MediaPlayer&)>&& callback) mutable {
-            this->performWithMediaPlayerOnMainThread(identifier, WTFMove(callback));
+        return IPC::ScopedActiveMessageReceiveQueue { RemoteGPU::create([this] (std::variant<WebCore::MediaPlayerIdentifier, WebKit::RemoteVideoFrameReference> identifier, Function<void(RefPtr<WebCore::VideoFrame>)>&& callback) mutable {
+            return WTF::switchOn(identifier, [&callback, this] (WebCore::MediaPlayerIdentifier i) {
```
you're already capture by reference the line above, may as well continue, it's shorter/neater and will capture the same thing

jyavenard commented on Source/WebKit/WebProcess/GPU/media/RemoteVideoFrameProxy.cpp:
```diff
@@ -127,6 +127,14 @@ CVPixelBufferRef RemoteVideoFrameProxy::pixelBuffer() const
+Vector<uint8_t> RemoteVideoFrameProxy::resourceIdentifierVector() const
+{
+    auto ref = newReadReference();
+    Vector<uint8_t> result(sizeof(ref));
```
Vector<uint8_t> result { reinterpret_cast< uint8_t*>(&ref), sizeof(ref) };

mwyrzykowski replied:
I got rid of the memcpy / Vector entirely :)

jyavenard commented on Source/WebCore/PAL/pal/graphics/WebGPU/WebGPUExternalTextureDescriptor.h:
```diff
@@ -27,13 +27,31 @@
 
+struct HTMLVideoElementIdentifier {
+    uint64_t value;
+    HTMLVideoElementIdentifier(uint64_t v)
```
you don't need to define a constructor here.

jyavenard commented on Source/WebCore/PAL/pal/graphics/WebGPU/WebGPUExternalTextureDescriptor.h:
```diff
@@ -27,13 +27,31 @@
+};
+struct WebCodecsVideoFrameIdentifier {
+    Vector<uint8_t> value;
+    WebCodecsVideoFrameIdentifier(Vector<uint8_t>&& v)
```
nor here, unless you want an implicit constructor

jyavenard commented on Source/WebKit/GPUProcess/GPUConnectionToWebProcess.cpp:
```diff
@@ -647,8 +647,14 @@ void GPUConnectionToWebProcess::createRemoteGPU(WebGPUIdentifier identifier, Ren
+                    callback(player.videoFrameForCurrentTime());
+                });
+            }, [&callback, this] (WebKit::RemoteVideoFrameReference i) {
+                callback(this->videoFrameObjectHeap().get(RemoteVideoFrameReadReference(WTFMove(i))));
```
if you capture by reference this is captured by default, don't need the `this->`

jyavenard commented on Source/WebCore/Modules/WebGPU/GPUExternalTextureDescriptor.h:
```diff
@@ -28,25 +28,51 @@
+#if ENABLE(WEB_CODECS)
+        return WTF::switchOn(videoSource, [] (const RefPtr<HTMLVideoElement> videoElement) -> PAL::WebGPU::VideoSourceIdentifier {
+            auto playerIdentifier = videoElement->playerIdentifier();
+            return PAL::WebGPU::HTMLVideoElementIdentifier { playerIdentifier.has_value() ? playerIdentifier->toUInt64() : 0 };
```
personally, I like simply testing `playerIndentifier ? playerIndentifer-> ... `

jyavenard commented on Source/WebKit/GPUProcess/GPUConnectionToWebProcess.cpp:
```diff
@@ -647,8 +647,14 @@ void GPUConnectionToWebProcess::createRemoteGPU(WebGPUIdentifier identifier, Ren
-            this->performWithMediaPlayerOnMainThread(identifier, WTFMove(callback));
+        return IPC::ScopedActiveMessageReceiveQueue { RemoteGPU::create([&] (std::variant<WebCore::MediaPlayerIdentifier, WebKit::RemoteVideoFrameReference> identifier, Function<void(RefPtr<WebCore::VideoFrame>)>&& callback) mutable {
+            return WTF::switchOn(identifier, [&] (WebCore::MediaPlayerIdentifier i) {
+                performWithMediaPlayerOnMainThread(i, [&callback](auto& player) {
```
you can't capture by reference here, callback would have gone out of scope prior. Capture with [callback = WTFMove(callback)]

jyavenard replied:
And is the call back to `capture()` thread safe? What is GPUConnectToWebProcess thread ?

mwyrzykowski replied:
@jyavenard is it really an issue to capture a reference here? `performWithMediaPlayerOnMainThread` is synchronous and the callback runs prior to `performWithMediaPlayerOnMainThread` returning. It uses a semaphore and blocks the calling thread until the callback runs. The semaphore is signaled after the callback completes. So the lifetime of `callback` should not expire prior to it being used here.

But maybe I should not be relying on the implementation details of `performWithMediaPlayerOnMainThread` and perform the copy as you suggest instead? What do you think?

mwyrzykowski replied:
Regarding the thread safety, since `performWithMediaPlayerOnMainThread` uses a BinarySemaphore, I think there are no thread safety considerations. 

However again this is relying on the implementation of `performWithMediaPlayerOnMainThread`. But I don't think `performWithMediaPlayerOnMainThread` thread safety guarantees could be changed at this point without breaking all callers who may rely on the current implementation.

litherum commented on Source/WebCore/PAL/pal/graphics/WebGPU/WebGPUExternalTextureDescriptor.h:
```diff
@@ -27,15 +27,29 @@
+#if PLATFORM(COCOA)
+    RetainPtr<CVPixelBufferRef> pixelBuffer { nullptr };
+#else
     CVPixelBufferRef pixelBuffer { nullptr };
```
This doesn't make a lot of sense. Why should there be be a `CVFoo` type in `!PLATFORM(COCOA)`?

mwyrzykowski replied:
It is to make WinCairo compile without adding `#if ENABLE(VIDEO) && PLATFORM(COCOA)` in a few places.

But indeed it should probably not be here. I will add `#if ENABLE(VIDEO) && PLATFORM(COCOA)` in a few places instead

litherum commented on Source/WebGPU/WebGPU/BindGroup.mm:
```diff
@@ -78,13 +78,280 @@ static MTLRenderStages metalRenderStage(ShaderStage shaderStage)
+        return plane ? MTLPixelFormatRG8Unorm : MTLPixelFormatR8Unorm;
+    };
+
+    switch (pixelFormat) {
```
I'm going to assume that this big table is right

Or at least, if it isn't, rigorous testing will discover any incorrect parts

litherum commented on Source/WebGPU/WebGPU/BindGroup.mm:
```diff
@@ -78,13 +78,280 @@ static MTLRenderStages metalRenderStage(ShaderStage shaderStage)
+static MTLPixelFormat metalPixelFormat(CVPixelBufferRef pixelBuffer, size_t plane)
+{
+    auto pixelFormat = CVPixelBufferGetPixelFormatType(pixelBuffer);
+    auto biplanarFormat = ^(int plane) {
```
Can we use C++ lambdas instead of blocks?

Blocks are useful for passing between framework boundaries, but if we're entirely within a single framework, we should prefer C++ lambas

mwyrzykowski replied:
I can make it a lambda, but what is the difference between C++ lambdas and blocks from clang's perspective? I thought they were basically the same.

litherum replied:
Oh, the difference is just being more idiomatic, rather than performance or anything like that.

mwyrzykowski replied:
Sounds good, I'll switch it to a lambda 👍

litherum commented on Source/WebGPU/WebGPU/BindGroup.mm:
```diff
@@ -78,13 +78,280 @@ static MTLRenderStages metalRenderStage(ShaderStage shaderStage)
 #if HAVE(COREVIDEO_METAL_SUPPORT)
     UNUSED_PARAM(colorSpace);
 
+    if (!CVPixelBufferGetIOSurface(pixelBuffer)) {
```
what happens if `CVPixelBufferGetIOSurface(pixelBuffer)` returns true? Surely we should have an `else` block below?

mwyrzykowski replied:
There is a path below. That is the existing GPU path. The `else` is omitted due to early return

litherum commented on Source/WebKit/GPUProcess/graphics/WebGPU/RemoteDevice.cpp:
```diff
@@ -152,24 +153,61 @@ void RemoteDevice::createSampler(const WebGPU::SamplerDescriptor& descriptor, We
+#if ENABLE(VIDEO) && PLATFORM(COCOA)
+void RemoteDevice::setSharedVideoFrameSemaphore(IPC::Semaphore&& semaphore)
+{
+    m_sharedVideoFrameReader.setSemaphore(WTFMove(semaphore));
```
Wow, really? We need semaphores to use external textures? Why?

mwyrzykowski replied:
This is how `SharedVideoFrameReader / Writer` is implemented. It is similar to what is done in `RemoteGraphicsContextGL`

Note we only encounter this path for local memory external textures. For external textures decoded in the GPU Process, we will not encounter this path.

litherum commented on Source/WebKit/WebProcess/GPU/media/RemoteVideoFrameProxy.h:
```diff
@@ -90,6 +90,7 @@ class RemoteVideoFrameProxy final : public WebCore::VideoFrame {
     bool isRemoteProxy() const final { return true; }
 #if PLATFORM(COCOA)
     CVPixelBufferRef pixelBuffer() const final;
+    std::pair<uint64_t, uint64_t> resourceIdentifier() const final;
```
This looks like it could use a typedef

mwyrzykowski committed:
[WebGPU] Support VideoFrame as the source of GPUExternalTexture
https://bugs.webkit.org/show_bug.cgi?id=256875
<radar://109442655>

Reviewed by Myles C. Maxfield.

Support VideoFrame as the source of a GPUExternalTexture.

* Source/WebCore/Modules/WebGPU/GPUExternalTextureDescriptor.h:
(WebCore::GPUExternalTextureDescriptor::mediaIdentifierForSource):
(WebCore::GPUExternalTextureDescriptor::convertToBacking const):
* Source/WebCore/Modules/WebGPU/GPUExternalTextureDescriptor.idl:
* Source/WebCore/PAL/pal/graphics/WebGPU/WebGPUExternalTextureDescriptor.h:
(PAL::WebGPU::HTMLVideoElementIdentifier::HTMLVideoElementIdentifier):
(PAL::WebGPU::WebCodecsVideoFrameIdentifier::WebCodecsVideoFrameIdentifier):
Support either HTMLVideoElement or WebCodecsVideoFrame as the source of a GPUExternalTexture.

* Source/WebCore/platform/VideoFrame.h:
(WebCore::VideoFrame::resourceIdentifierVector const):
Default implementation returns and empty byte vector.

* Source/WebCore/platform/VideoPixelFormat.cpp:
(WebCore::convertVideoFramePixelFormat):
Allow kCVPixelFormatType_Lossless_420YpCbCr8BiPlanarVideoRange formats.

* Source/WebGPU/WebGPU/ExternalTexture.h:
(WebGPU::ExternalTexture::pixelBuffer const):
Retain the CVPixelBuffer which wasn't previously occurring.

* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.cpp:
(WebKit::GPUConnectionToWebProcess::createRemoteGPU):
Update the lambda to work with HTMLVideoElements and VideoFrames as inputs.

* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteDevice.cpp:
(WebKit::populateConvertedDescriptor):
(WebKit::RemoteDevice::importExternalTexture):
Support HTMLVideoElements and VideoFrames as inputs.

* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteDevice.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteGPU.h:

* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/WebGPU/WebGPUExternalTextureDescriptor.cpp:
(WebKit::WebGPU::ConvertToBackingContext::convertToBacking):
(WebKit::WebGPU::ConvertFromBackingContext::convertFromBacking):
PAL and WebCore representations are nearly identical.

* Source/WebKit/Shared/WebGPU/WebGPUExternalTextureDescriptor.h:
* Source/WebKit/Shared/WebGPU/WebGPUExternalTextureDescriptor.serialization.in:
* Source/WebKit/WebProcess/GPU/media/RemoteVideoFrameProxy.cpp:
(WebKit::RemoteVideoFrameProxy::resourceIdentifierVector const):
Encode the readRefrence into a byte-vector.

* Source/WebKit/WebProcess/GPU/media/RemoteVideoFrameProxy.h:

Canonical link: https://commits.webkit.org/264847@main

youennf commented on Source/WebCore/Modules/WebGPU/GPUExternalTextureDescriptor.h:
```diff
@@ -28,25 +28,64 @@
 
 struct GPUExternalTextureDescriptor : public GPUObjectDescriptorBase {
+
+    static PAL::WebGPU::VideoSourceIdentifier mediaIdentifierForSource(const GPUVideoSource& videoSource, CVPixelBufferRef& outPixelBuffer)
```
We try to limit out parameters.
Should we instead return a pair here? That would ensure outPixelBuffer is always either nullptr or a real value, not an uninitialised value.

youennf commented on Source/WebKit/WebProcess/GPU/media/RemoteVideoFrameProxy.cpp:
```diff
@@ -127,6 +127,12 @@ CVPixelBufferRef RemoteVideoFrameProxy::pixelBuffer() const
+
+VideoFrame::ResourceIdentifier RemoteVideoFrameProxy::resourceIdentifier() const
+{
+    auto ref = newReadReference();
```
This is potentially dangerous as creating a read reference increments the pending read counter.
When creating such a read reference, we should ensure to send the read reference to GPUProcess through IPC.
It is difficult to me to tell whether the new code path actually ensures this.
In any case, resourceIdentifier() makes it as if it is without side effect, which is not the case.
We should at least find a better name for this routine. If we use SharedVideoFrame and a single IPC message, we might be able to remove this routine.

This makes me wonder whether we should not revisit the current remote video frame strategy, given this and other stuff (some desire to create some remote video frame in UIProcess, desire to move some video frames from one web process to another).

mwyrzykowski replied:
Maybe I can use `SharedVideoFrameWriter` instead since I think this is achieving the same thing, I will attempt to refactor this.

youennf commented on Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteDeviceProxy.cpp:
```diff
@@ -128,8 +129,28 @@ Ref<PAL::WebGPU::ExternalTexture> RemoteDeviceProxy::importExternalTexture(const
+        auto sendResult = send(Messages::RemoteDevice::ImportExternalTextureFromPixelBuffer(*convertedDescriptor, sharedVideoFrameBuffer, identifier));
+        UNUSED_VARIABLE(sendResult);
+    }  else {
+        auto sendResult = send(Messages::RemoteDevice::ImportExternalTexture(*convertedDescriptor, identifier));
```
This is a strange pattern, SharedVideoFrame is supposed to handle all video frames, remote or not.
Why are we separating them? Usually we would send a single SharedVideoFrame object.

Also, creating a SharedVideoFrame for each video frame will mean that we will create a shared memory for each video frame, which is inefficient in case we do repetitive GPU processing of video frames of the same type/same size (say a MediaStreamTrack).

Also, I do not see any test for any of either the remote or local code path.

mwyrzykowski replied:
I think I can use `SharedVideoFrame` as long as I pass the VideoFrame as an opaque pointer through the WebGPU PAL layer, which can not see WebCore types. I will try that instead.

mwyrzykowski replied:
Regarding testing, we have some tests from the WebGPU CTS checked in at `./LayoutTests/http/tests/webgpu/cts.https.html`:
```
<meta name=variant content='webgpu:web_platform,external_texture,video:importExternalTexture,sample:*'>
<meta name=variant content='webgpu:web_platform,external_texture,video:importExternalTexture,expired:*'>
<meta name=variant content='webgpu:web_platform,external_texture,video:importExternalTexture,compute:*'>
```

these tests are not yet passing with these changes. I'm investigating why as part of https://bugs.webkit.org/show_bug.cgi?id=257724

youennf commented on Source/WebCore/Modules/WebGPU/GPUExternalTextureDescriptor.h:
```diff
@@ -28,25 +28,64 @@
+            return PAL::WebGPU::HTMLVideoElementIdentifier { playerIdentifier ? playerIdentifier->toUInt64() : 0 };
+        }
+        , [&outPixelBuffer] (const RefPtr<WebCodecsVideoFrame> videoFrame) -> PAL::WebGPU::VideoSourceIdentifier {
+#if PLATFORM(COCOA)
```
Why is it COCOA specific?

mwyrzykowski replied:
Oh it doesn't need to be COCOA specific, only because it is calling `pixelBuffer`. But I think I can remove this by passing an opaque pointer through the PAL layer and working with VideoFrame instances directly.

youennf reviewed (commented):
Some thoughts below
